### PR TITLE
feat: support custom onnx file and update model signatures

### PR DESCRIPTION
**Pull request by ZiniuYu**
This PR allows the user to set filepath for pretrained custom onnx model and unify the signature of tensorrt, onnx, and huggingface clip runtime.
We also add md5 verification to make sure users download the latest and correct models.
Since tensorrt models are extremely picky for many dependencies, we no longer support tensorrt model hosting, which means that it will first download the latest onnx model and convert to trt during runtime and save for it later use.

TODO:
- [x] Add parameters to allow user select custom pretrained onnx model
- [x] Convert existing models to have updated signatures
- [x] Modify other executors to be compatible with the new structure
- [x] Add MD5 verification for downloaded model

### CODE REVIEW

File: server/clip_server/model/clip_onnx.py [hunk 1]
```diff
@@ -17,8 +17,11 @@
 
 
 class CLIPOnnxModel:
-    def __init__(self, name: str = None):
-        if name in _MODELS:
+    def __init__(self, name: str = None, model_paths: str = None):
+        if model_paths is not None:
```
Comment (numb3r3) [inline server/clip_server/model/clip_onnx.py:21]: Valid the `model_paths`, raise warning to tell user what's the correct input

File: server/clip_server/executors/clip_onnx.py [hunk 2]
```diff
@@ -23,6 +23,7 @@ def __init__(
         num_worker_preprocess: int = 4,
         minibatch_size: int = 32,
         traversal_paths: str = '@r',
+        model_paths: Optional[str] = None,
```
Comment (numb3r3) [inline server/clip_server/executors/clip_onnx.py:26]: the param `model_paths` looks like a list of values. Let's use `model_path`

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip_onnx.py (hunk 1 above)
Comment (numb3r3) [inline server/clip_server/model/clip_onnx.py:21]: The `name` param should always be one of the `_MODELS` even if the given `model_paths` is not empty.

File: server/clip_server/model/clip_onnx.py [hunk 3]
```diff
@@ -17,15 +17,30 @@
...
+                cache_dir = os.path.expanduser(
+                    f'~/.cache/clip/{name.replace("/", "-")}'
+                )
+                self._textual_path = _download(
+                    _S3_BUCKET + _MODELS[name][0], cache_dir, with_resume=True
+                )
+                self._visual_path = _download(
+                    _S3_BUCKET + _MODELS[name][1], cache_dir, with_resume=True
+                )
         else:
```
Comment (numb3r3) [inline server/clip_server/model/clip_onnx.py:44]: Polish the code

File: tests/conftest.py [hunk 4]
```diff
@@ -16,16 +16,26 @@ def random_port():
...
-        from clip_server.executors.clip_hg import CLIPEncoder
+        import os
+        from clip_server.executors.clip_onnx import CLIPEncoder
 
-    f = Flow(port=port_generator()).add(name=request.param, uses=CLIPEncoder)
+        f = Flow(port=port_generator()).add(
+            name=request.param,
+            uses=CLIPEncoder,
+            uses_with={'model_path': os.path.expanduser('~/.cache/clip/ViT-B-32')},
+        )
```
Comment (numb3r3) [inline tests/conftest.py:38]: With this changes,  `request.param == torch` also use `clip_onnx`

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip_onnx.py (hunk 3 above)
Comment (numb3r3) [reply server/clip_server/model/clip_onnx.py]:
```

class CLIPOnnxModel:
    def __init__(self, name: str = None, model_path: str = None):
        if name in _MODELS:
            if not model_path:
                cache_dir = os.path.expanduser(
                    f'~/.cache/clip/{name.replace("/", "-")}'
                )
                self._textual_path = _download(
                    _S3_BUCKET + _MODELS[name][0], cache_dir, with_resume=True
                )
                self._visual_path = _download(
                    _S3_BUCKET + _MODELS[name][1], cache_dir, with_resume=True
                )
            elif os.path.isdir(model_path):
                self._textual_path = os.path.join(model_path, 'textual.onnx')
                self._visual_path = os.path.join(model_path, 'visual.onnx')
                if not os.path.isfile(self._textual_path) or os.path.isfile(self._visual_path):
                    raise ValueError(f'{model_path} does not contain `textual.onnx` and `visual.onnx`')
        else:
            raise RuntimeError(
                f'Model {name} not found; available models = {available_models()}'
            )
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip.py [hunk 5]
```diff
@@ -335,6 +336,7 @@ def tokenize(
     eot_token = _tokenizer.encoder['<|endoftext|>']
     all_tokens = [[sot_token] + _tokenizer.encode(text) + [eot_token] for text in texts]
     result = torch.zeros(len(all_tokens), context_length, dtype=torch.long)
+    attention_masks = torch.zeros(len(all_tokens), context_length, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (ZiniuYu) [reply server/clip_server/model/clip.py]: this is blocked

File: server/clip_server/model/clip_trt.py [hunk 6]
```diff
@@ -34,77 +35,77 @@ def __init__(
...
-                fp16=False,
-                int8=False,
-            )
-            save_engine(text_engine, self._textual_path)
-
         self._textual_engine = load_engine(runtime, self._textual_path)
         self._visual_engine = load_engine(runtime, self._visual_path)
 
     def encode_image(self, onnx_image):
         (visual_output,) = self._visual_engine({'input': onnx_image})
```
Comment (numb3r3) [inline server/clip_server/model/clip_trt.py:106]: The visual input name should be "pixels"?

Comment (numb3r3) [reply server/clip_server/model/clip_trt.py:106]: And we can directly use `onnx_image` as it is already an dict.

File: server/clip_server/model/clip.py [hunk 7]
```diff
@@ -50,12 +51,20 @@
 }
 
 
-def _download(url: str, root: str, with_resume: bool = True):
+def _download(
+    url: str, md5: str, root: str, with_resume: bool = True, max_attempts: int = 3
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:55]: `md5` should also be optional. If md5 is not provided, then the checking step is ignored

File: server/clip_server/model/clip.py [hunk 8]
```diff
@@ -127,6 +136,11 @@ def _download(url: str, root: str, with_resume: bool = True):
             ):
                 shutil.move(tmp_file_path, download_target)
 
+    if hashlib.md5(open(download_target, 'rb').read()).hexdigest() != md5:
+        print(f'MD5 mismatch for {download_target}, will retry')
+        os.remove(download_target)
+        return _download(url, md5, root, with_resume, max_attempts - 1)
+
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:143]: this step should be placed before `shutil.move`?

File: server/clip_server/model/clip.py [hunk 9]
```diff
@@ -309,7 +324,7 @@ def patch_float(module):
 
 def tokenize(
     texts: Union[str, List[str]], context_length: int = 77, truncate: bool = True
-) -> torch.LongTensor:
+) -> {}:
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:327]: The return type should be `Dict`, not `{}`

File: server/clip_server/model/clip.py [hunk 10]
```diff
@@ -326,15 +341,23 @@ def tokenize(
...
     sot_token = _tokenizer.encoder['<|startoftext|>']
     eot_token = _tokenizer.encoder['<|endoftext|>']
     all_tokens = [[sot_token] + _tokenizer.encode(text) + [eot_token] for text in texts]
-    result = torch.zeros(len(all_tokens), context_length, dtype=torch.long)
+    # input_ids = torch.zeros(
+    #     len(all_tokens), min(len(max(all_tokens, key=len)), context_length), dtype=torch.long
+    # )
+    # attention_masks = torch.zeros(
+    #     len(all_tokens), min(len(max(all_tokens, key=len)), context_length), dtype=torch.long
+    # )
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:358]: please clean the unused codes

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip.py [hunk 11]
```diff
@@ -205,8 +225,9 @@ def load(
     """
     if name in _MODELS:
         model_path = _download(
```
[the remark this answers is not in the document]
Comment (numb3r3) [reply server/clip_server/model/clip.py:227]: And let's pass kwargs style parameters, `foo(.., root='root/path', md5='xxx', ...)`. So that the code is more easy to follow

File: server/clip_server/model/clip.py [hunk 12]
```diff
@@ -326,15 +347,18 @@ def tokenize(
 
     Returns
     -------
-    A two-dimensional tensor containing the resulting tokens, shape = [number of input strings, context_length]
+    A dict of tokenized representations of the input strings and their corresponding attention masks with both
+        shape = [number of input strings, context_length]
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:351]: shape = [batch size, context_length]

File: server/clip_server/tensorrt-flow.yml [hunk 13]
```diff
@@ -9,4 +9,5 @@ executors:
       metas:
         py_modules:
           - executors/clip_tensorrt.py
+      timeout_ready: 6000000
```
Comment (numb3r3) [inline server/clip_server/tensorrt-flow.yml:12]: Please double check what's the meaning of `6000000`? If I'm correct, the default value is 6 min. let's change it to 15 min (or need more time?).

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/tensorrt-flow.yml (hunk 13 above)
Comment (ZiniuYu) [reply server/clip_server/tensorrt-flow.yml:12]: In bj gpu sometimes it needs 20-30min, the default value is 600,000 https://docs.jina.ai/fundamentals/flow/yaml-spec/#executors

File: server/clip_server/model/clip.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ZiniuYu) [reply server/clip_server/model/clip.py:227]: 👌

File: server/clip_server/model/clip_trt.py [hunk 14]
```diff
@@ -14,18 +14,19 @@
...
+_MODELS = [
+    'RN50',
+    'RN101',
+    'RN50x4',
+    # 'RN50x16',
+    # 'RN50x64',
+    'ViT-B/32',
+    'ViT-B/16',
+    'ViT-L/14',
+    # 'ViT-L/14@336px',
```
[the remark this answers is not in the document]
Comment (ZiniuYu) [reply server/clip_server/model/clip_trt.py:28]: I reran some tests to reconvert the model, not only did it take ~30min to download and convert, but also met this error 
` [TRT] [E] 10: [optimizer.cpp::computeCosts::3628] Error Code 10: Internal Error (Could not find any implementation for node {ForeignNode[onnx::MatMul_3417 + (Unnamed Layer* 87) [Shuffle]...MatMul_3011]}.)`
 Looks like it might be some problem with trt itself -> https://github.com/NVIDIA/TensorRT/issues/2124
 I would suggest temporarily removing the support of VL14 model for trt and wait for several new releases from trt.

File: server/clip_server/tensorrt-flow.yml (hunk 13 above)
Comment (ZiniuYu) [reply server/clip_server/tensorrt-flow.yml]: np

File: server/clip_server/executors/helper.py [hunk 15]
```diff
@@ -33,34 +33,39 @@ def preproc_image(
...
 
     if return_np:
-        tensors_batch = tensors_batch.cpu().numpy()
+        inputs['pixel_values'] = inputs['pixel_values'].cpu().numpy()
     else:
-        tensors_batch = tensors_batch.to(device)
+        inputs['pixel_values'] = inputs['pixel_values'].to(device)
 
-    return da, tensors_batch
+    return da, inputs
```
[the remark this answers is not in the document]
Comment (ZiniuYu) [reply server/clip_server/executors/helper.py]: to match the structure of text_encode, just modified it anyway

File: server/clip_server/model/clip.py [hunk 16]
```diff
@@ -124,8 +136,16 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                and (
+                    not md5
+                    or hashlib.md5(open(tmp_file_path, 'rb').read()).hexdigest() == md5
+                )
             ):
                 shutil.move(tmp_file_path, download_target)
+            else:
+                print(f'MD5 mismatch for {download_target}, will retry')
+                os.remove(tmp_file_path)
+                return _download(url, root, md5, with_resume, max_attempts - 1)
```
[the remark this answers is not in the document]
Comment (ZiniuYu) [reply server/clip_server/model/clip.py]: np and done

File: server/clip_server/model/clip_trt.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (numb3r3) [reply server/clip_server/model/clip_trt.py:28]: I'm fine to remove VL14 support. But make sure to 
- update the documentation
- raise an issue to track this problem.

File: server/clip_server/model/clip.py [hunk 17]
```diff
@@ -81,53 +91,71 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                    with urllib.request.urlopen(req) as source:
+                        while True:
+                            buffer = source.read(8192)
+                            if not buffer:
+                                break
+
+                            output.write(buffer)
+                            progress.update(task, advance=len(buffer))
+            except Exception as ex:
+                raise ex
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:132]: In this case, we need to restart the download, rather than raise exception

File: server/clip_server/model/clip.py [hunk 18]
```diff
@@ -81,53 +91,71 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                else:
+                    progress.console.print(
+                        f'MD5 mismatch for {download_target}, maybe the file is not completely downloaded. '
+                        f'Retrying now...'
+                    )
+                    os.remove(tmp_file_path)
+                    progress.reset(task)
+
+    if os.path.isfile(download_target) and (
+        not md5 or hashlib.md5(open(download_target, 'rb').read()).hexdigest() == md5
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:154]: I suggest wrapping this validation as a helper function

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip.py [hunk 19]
```diff
@@ -81,53 +95,67 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                    with urllib.request.urlopen(req) as source:
+                        while True:
+                            buffer = source.read(8192)
+                            if not buffer:
+                                break
+
+                            output.write(buffer)
+                            progress.update(task, advance=len(buffer))
+            except Exception as ex:
+                progress.console.print(f'Error: {ex} Retrying now...')
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:132]: please remember to reset the progress bar.

File: server/clip_server/model/clip.py [hunk 20]
```diff
@@ -81,53 +95,67 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                continue
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:133]: The `continue` state cannot stop the execution of the `finally` block

File: server/clip_server/model/clip.py (hunk 19 above)
Comment (numb3r3) [reply server/clip_server/model/clip.py:132]: And what's more, the logging message is not correct. `Error: {ex} Retrying now...` looks confusing with exceeded retries.

File: server/clip_server/model/clip.py [hunk 21]
```diff
@@ -81,53 +95,67 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+
+            finally:
+                # rename the temp download file to the correct name if fully downloaded
+                if os.path.exists(tmp_file_path) and (
+                    total_bytes == os.path.getsize(tmp_file_path)
+                    and (
+                        not md5
+                        or hashlib.md5(open(tmp_file_path, 'rb').read()).hexdigest()
+                        == md5
+                    )
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:143]: use the `_validate` function instead

File: server/clip_server/model/clip.py [hunk 22]
```diff
@@ -81,53 +95,67 @@ def _download(url: str, root: str, with_resume: bool = True):
...
+                ):
+                    shutil.move(tmp_file_path, download_target)
+                    break
+                else:
+                    progress.console.print(
+                        f'MD5 mismatch for {download_target}, maybe the file is not completely downloaded. '
```
Comment (numb3r3) [inline server/clip_server/model/clip.py:149]: can we output the expected and real md5 hash code? This will help us to debug when the user has trouble downloading the model.

File: server/clip_server/model/clip_onnx.py [hunk 23]
```diff
@@ -2,30 +2,85 @@
...
+                )
+            else:
+                if os.path.isdir(model_path):
+                    self._textual_path = os.path.join(model_path, 'textual.onnx')
+                    self._visual_path = os.path.join(model_path, 'visual.onnx')
+                    if not os.path.isfile(self._textual_path) or not os.path.isfile(
+                        self._visual_path
+                    ):
+                        raise RuntimeError(
+                            f'{model_path} does not contain `textual.onnx` and `visual.onnx`'
```
Comment (numb3r3) [inline server/clip_server/model/clip_onnx.py:78]: clarify the logging f`The given model path {model_path} ....`

File: server/clip_server/model/clip_trt.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (numb3r3) [reply server/clip_server/model/clip_trt.py:28]: Please make sure to create a ticket to track this issue.

File: server/clip_server/model/clip_trt.py [hunk 24]
```diff
@@ -34,77 +35,77 @@ def __init__(
...
-            self._visual_path = _download(_S3_BUCKET + _MODELS[name][1], cache_dir)
+
+            self._textual_path = os.path.join(
+                cache_dir,
+                f'textual.{ONNX_MODELS[name][0][1]}.trt',
+            )
+            self._visual_path = os.path.join(
+                cache_dir,
+                f'visual.{ONNX_MODELS[name][1][1]}.trt',
+            )
```
Comment (numb3r3) [inline server/clip_server/model/clip_trt.py:48]: The local onnx model is not validated with MD5?

File: server/clip_server/model/clip_trt.py [hunk 25]
```diff
@@ -34,77 +35,77 @@ def __init__(
...
+
+            if not os.path.exists(self._textual_path) or not os.path.exists(
```
Comment (numb3r3) [inline server/clip_server/model/clip_trt.py:50]: We can separate the converting into two branches: virtual and textual model converting seperatedly.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/clip_trt.py (hunk 25 above)
Comment (ZiniuYu) [reply server/clip_server/model/clip_trt.py:50]: i think they should either succeed or fail simultaneously

Comment (numb3r3) [reply server/clip_server/model/clip_trt.py:50]: The motivation is to speed up the converting when the converting is stopped by `CTRL+C`. But I'm fine to keep the current implementation which is much safe.

File: server/clip_server/model/clip_trt.py (hunk 24 above)
Comment (ZiniuYu) [reply server/clip_server/model/clip_trt.py:48]: no they are the local trt files and are generated by us
there is no predefined md5 to validate with

Comment (numb3r3) [reply server/clip_server/model/clip_trt.py:48]: What I'm asking is actually the existing of onnx model file, not the TRT model. But anyway, I think the following `onnx_model = CLIPOnnxModel(self._name)` involves the downloading and validation procedure. Hence, there is no problem about this part.

**Review by numb3r3 — CHANGES_REQUESTED** [review]
